### PR TITLE
Fix TrideNode wildcard and regexp management

### PR DESCRIPTION
**Pull request by Wonshtrum**
Add unit test and e2e test for wildcard behavior.
Fix the TrieNode (is_empty, insert, remove, lookup_mut)
Documenting comments and variable renaming in the router module.

### CODE REVIEW

File: lib/src/router/mod.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl Router {
                 //FIXME: necessary ti build on stable rust (1.35), can be removed once 1.36 is there
                 let mut empty = true;
                 if let Some((_, ref mut paths)) =
-                    self.tree.domain_lookup_mut(hostname.as_bytes(), false)
+                    self.tree.domain_lookup_mut(hostname.as_bytes(), true)
```
Comment (Wonshtrum) [inline lib/src/router/mod.rs:205]: This is the actual fix.

**Review by Keksoj — APPROVED** [review]
looks very good to me!

**Comment by Wonshtrum** [comment]
ABORT! This doesn't work!

**Review by Wonshtrum — COMMENTED** [review]
ABORT! This doesn't work!

**Review by Keksoj — CHANGES_REQUESTED** [review]
well if it doesn't work, let's block the merge

**Comment by Wonshtrum** [comment]
The problem was more profound than previously thought.
TLDR: `lookup_mut` and `lookup` had the same behavior (apart from mutability), but are used very differently:
- `lookup` is only used to check "real complete" URLs against the trie tree.
- `lookup_mut` can be used with "pattern" URLs (wildcard and regex)

This PR changes `lookup_mut` to properly support wildcards and regexes in URLs as well as fixing some branches in `insert_recursive` and `remove_recursive` which didn't correctly propagate.

File: lib/src/router/mod.rs [hunk 2]
```diff
@@ -685,6 +685,112 @@ mod tests {
...
+            Some(Route::ClusterId("base".to_string()))
+        );
+        assert_eq!(
+            router.lookup("www.sozu.io".as_bytes(), "/api".as_bytes(), &Method::Get),
+            Some(Route::ClusterId("api".to_string()))
+        );
+    }
+
+    #[test]
+    fn multiple_children_with_wildcard() {
```
[the remark this answers is not in the document]
Comment (Wonshtrum) [reply lib/src/router/mod.rs:721]: The two tests are really similar in nature, the only difference is the position of the wildcard in the tree.

multiple_children_on_wildcard:
```
[com]
     \
      [.*]  <- this wildcard has multiple children
     /    \
[www]      [doc]
                \
                 [www]
```
multiple_children_with_wildcard:
```
             [com]  <- this node has multiple children with a wildcard
            /     \
      [sozu]       [.*]
     /                 \
[www]                   [www]
```
What would you change their names to?
Maybe multiple_children_with_wildcard -> wildcard_among_multiple_children?

[the remark this answers is not in the document]
Comment (Keksoj) [reply lib/src/router/mod.rs:721]: how about:
1. `multiple_children_to_a_wildcard`
2.`multiple_children_including_one_with_wildcard`

File: lib/src/router/pattern_trie.rs [hunk 3]
```diff
@@ -229,18 +254,16 @@ impl<V: Debug + Clone> TrieNode<V> {
             Some(child) => match child.remove_recursive(prefix) {
                 RemoveResult::NotFound => return RemoveResult::NotFound,
                 RemoveResult::Ok => {
-                    if !child.is_empty() {
-                        return RemoveResult::Ok;
+                    if child.is_empty() {
+                        println!("child of {suffix:?} is empty:");
```
Comment (FlorentinDUBOIS) [inline lib/src/router/pattern_trie.rs:258]: Could you remove this println and the one below

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]

**Review by FlorentinDUBOIS — APPROVED** [review]

**Review by FlorentinDUBOIS — APPROVED** [review]
